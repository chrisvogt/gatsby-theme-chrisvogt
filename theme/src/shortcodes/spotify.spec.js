import React from 'react'
import renderer from 'react-test-renderer'
import Spotify from './spotify'

// Mock fetch globally
global.fetch = jest.fn()

describe('Spotify', () => {
  beforeEach(() => {
    fetch.mockClear()
  })

  it('renders loading state initially', () => {
    const component = renderer.create(<Spotify spotifyURL='https://open.spotify.com/track/123' />)
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('renders error state when fetch fails', async () => {
    fetch.mockRejectedValueOnce(new Error('Network error'))

    let component
    await renderer.act(async () => {
      component = renderer.create(<Spotify spotifyURL='https://open.spotify.com/track/123' />)
    })

    // Wait for the async operation to complete
    await new Promise(resolve => setTimeout(resolve, 0))

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('renders error state when HTTP response is not ok', async () => {
    fetch.mockResolvedValueOnce({
      ok: false,
      status: 404
    })

    let component
    await renderer.act(async () => {
      component = renderer.create(<Spotify spotifyURL='https://open.spotify.com/track/123' />)
    })

    // Wait for the async operation to complete
    await new Promise(resolve => setTimeout(resolve, 0))

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('renders embed when fetch succeeds', async () => {
    const mockResponse = {
      html: '<iframe src="https://open.spotify.com/embed/track/123"></iframe>'
    }
    fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => mockResponse
    })

    let component
    await renderer.act(async () => {
      component = renderer.create(<Spotify spotifyURL='https://open.spotify.com/track/123' />)
    })

    // Wait for the async operation to complete
    await new Promise(resolve => setTimeout(resolve, 0))

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('does not render when no spotifyURL is provided', async () => {
    let component
    await renderer.act(async () => {
      component = renderer.create(<Spotify />)
    })

    // Wait for the async operation to complete
    await new Promise(resolve => setTimeout(resolve, 0))

    expect(component.toJSON()).toBeNull()
  })
})
