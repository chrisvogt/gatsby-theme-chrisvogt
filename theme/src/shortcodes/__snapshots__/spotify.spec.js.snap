// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Spotify renders embed when fetch succeeds 1`] = `
<div
  className="css-tmbyjz"
  dangerouslySetInnerHTML={
    {
      "__html": "<iframe src="https://open.spotify.com/embed/track/123"></iframe>",
    }
  }
/>
`;

exports[`Spotify renders error state when HTTP response is not ok 1`] = `
<div
  className="css-12dtzxe"
>
  Failed to load Spotify embed
</div>
`;

exports[`Spotify renders error state when fetch fails 1`] = `
<div
  className="css-12dtzxe"
>
  Failed to load Spotify embed
</div>
`;

exports[`Spotify renders loading state initially 1`] = `
<div
  className="css-12dtzxe"
>
  Loading...
</div>
`;
